// redis-om/dist/search/results-converter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.JsonSearchResultsConverter = exports.HashSearchResultsConverter = exports.SearchResultsConverter = void 0;
const converter_1 = require("../repository/converter");
class SearchResultsConverter {
    constructor(schema, results) {
        this.schema = schema;
        this.results = results;
    }
    get count() {
        let [count] = this.results;
        return Number.parseInt(count);
    }
    get ids() {
        return this.keys.map(key => key.replace(/^.*:/, ""));
    }
    get keys() {
        let [_count, ...keysAndValues] = this.results;
        return keysAndValues.filter((_entry, index) => index % 2 === 0);
    }
    get values() {
        let [_count, ...keysAndValues] = this.results;
        return keysAndValues.filter((_entry, index) => index % 2 !== 0);
    }
    get entities() {
        let ids = this.ids;
        let values = this.values;
        return values.map((array, index) => this.arrayToEntity(ids[index], array));
    }
}
exports.SearchResultsConverter = SearchResultsConverter;
class HashSearchResultsConverter extends SearchResultsConverter {
    arrayToEntity(id, array) {
        let keys = array.filter((_entry, index) => index % 2 === 0);
        let values = array.filter((_entry, index) => index % 2 !== 0);
        let hashData = keys.reduce((object, key, index) => {
            object[key] = values[index];
            return object;
        }, {});
        let converter = new converter_1.HashConverter(this.schema.definition);
        let entityData = converter.toEntityData(hashData);
        return new this.schema.entityCtor(this.schema, id, entityData);
    }
}
exports.HashSearchResultsConverter = HashSearchResultsConverter;
class JsonSearchResultsConverter extends SearchResultsConverter {
    arrayToEntity(id, array) {
        let index = array.findIndex(value => value === '$') + 1;
        let jsonString = array[index];
        let jsonData = JSON.parse(jsonString);
        let converter = new converter_1.JsonConverter(this.schema.definition);
        let entityData = converter.toEntityData(jsonData);
        return new this.schema.entityCtor(this.schema, id, entityData);
    }
}
exports.JsonSearchResultsConverter = JsonSearchResultsConverter;
