// redis-om/dist/search/where-date.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const where_field_1 = __importDefault(require("./where-field"));
class WhereDate extends where_field_1.default {
    constructor() {
        super(...arguments);
        this.lower = Number.NEGATIVE_INFINITY;
        this.upper = Number.POSITIVE_INFINITY;
        this.lowerExclusive = false;
        this.upperExclusive = false;
    }
    eq(value) {
        let epoch = this.coerceDateToEpoch(value);
        this.lower = epoch;
        this.upper = epoch;
        return this.search;
    }
    gt(value) {
        let epoch = this.coerceDateToEpoch(value);
        this.lower = epoch;
        this.lowerExclusive = true;
        return this.search;
    }
    gte(value) {
        this.lower = this.coerceDateToEpoch(value);
        return this.search;
    }
    lt(value) {
        this.upper = this.coerceDateToEpoch(value);
        this.upperExclusive = true;
        return this.search;
    }
    lte(value) {
        this.upper = this.coerceDateToEpoch(value);
        return this.search;
    }
    between(lower, upper) {
        this.lower = this.coerceDateToEpoch(lower);
        this.upper = this.coerceDateToEpoch(upper);
        return this.search;
    }
    equal(value) { return this.eq(value); }
    equals(value) { return this.eq(value); }
    equalTo(value) { return this.eq(value); }
    greaterThan(value) { return this.gt(value); }
    greaterThanOrEqualTo(value) { return this.gte(value); }
    lessThan(value) { return this.lt(value); }
    lessThanOrEqualTo(value) { return this.lte(value); }
    on(value) { return this.eq(value); }
    after(value) { return this.gt(value); }
    before(value) { return this.lt(value); }
    onOrAfter(value) { return this.gte(value); }
    onOrBefore(value) { return this.lte(value); }
    toString() {
        let lower = this.makeLowerString();
        let upper = this.makeUpperString();
        return this.buildQuery(`[${lower} ${upper}]`);
    }
    makeLowerString() {
        if (this.lower === Number.NEGATIVE_INFINITY)
            return '-inf';
        if (this.lowerExclusive)
            return `(${this.lower}`;
        return this.lower.toString();
    }
    makeUpperString() {
        if (this.upper === Number.POSITIVE_INFINITY)
            return '+inf';
        if (this.upperExclusive)
            return `(${this.upper}`;
        return this.upper.toString();
    }
    coerceDateToEpoch(value) {
        if (value instanceof Date)
            return value.getTime();
        if (typeof (value) === 'string')
            return new Date(value).getTime();
        return value;
    }
}
exports.default = WhereDate;
