// redis-om/dist/shims/redis-shim.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const redis_1 = require("redis");
const errors_1 = __importDefault(require("../errors"));
class RedisShim {
    constructor(urlOrConnection) {
        if (typeof (urlOrConnection) === 'string') {
            this.redis = (0, redis_1.createClient)({ url: urlOrConnection });
        }
        else {
            this.redis = urlOrConnection;
        }
    }
    async open() {
        await this.redis.connect();
    }
    async close() {
        await this.redis.quit();
    }
    execute(command) {
        return this.redis.sendCommand(command);
    }
    async unlink(key) {
        await this.redis.unlink(key);
    }
    async expire(key, ttl) {
        await this.redis.expire(key, ttl);
    }
    get(key) {
        return this.redis.get(key);
    }
    set(key, value) {
        return this.redis.set(key, value);
    }
    hgetall(key) {
        return this.redis.hGetAll(key);
    }
    async hsetall(key, data) {
        try {
            await this.redis.executeIsolated(async (isolatedClient) => {
                await isolatedClient.watch(key);
                await isolatedClient
                    .multi()
                    .unlink(key)
                    .hSet(key, data)
                    .exec();
            });
        }
        catch (error) {
            if (error.name === 'WatchError')
                throw new errors_1.default("Watch error when setting HASH.");
            throw error;
        }
    }
}
exports.default = RedisShim;
